src/gesture.js

```diff
@@ -297,7 +297,8 @@ export class Gestures {
         continue;
       }
       this.recognizers_[i].onTouchEnd(event);
-      if (!this.pending_[i] || this.pending_[i] < now) {
+      if ((!this.pending_[i] || this.pending_[i] < now)
+        && this.eventing_ != this.recognizers_[i]) {
```

Comment:
Perhaps pull this out into a variable to make it more clear. I think having a lot of complicated logic in an if-statement can get hard to follow. Something like:

```javascript
const isEventing = this.eventing_ == this.recognizers_[i];
if ((!this.pending_[i] || this.pending_[i] < now) && !isEventing) {
  this.stopTracking_(i);
}
```

or better one step further:

```javascript
const isEventing = this.eventing_ == this.recognizers_[i];
const isExpiredAfterEvent = !this.pending_[i] || this.pending_[i] < now;
if ((isExpiredAfterEvent && !isEventing) {
  this.stopTracking_(i);
}
```

I think this brings up a question for me: is it a problem that `this.pending_[i]` is < `now`? It seems like at this point, if another touchend were to happen (lets say you need 2 touches, you have 4, then you went to 3, then immediately to 2), then it seems like the if-statement above would call `stopTracking()`, even though you still have two touches.

This seems to indicate that a similar isEventing check might be needed above.

Reply:
Per discussion offline, pulled out the booleans into separate named constants. The previous block should not have an eventing check because of the `continue`. If it's eventing, then we definitely want to call the recognizer's `onTouchEnd`, not skip it. I think that would be a larger refactor that would be the scope for a different PR.